src/core/net/srp_client.cpp

```diff
@@ -1656,6 +1714,90 @@ void Client::ProcessAutoStart(void)
+void Client::SelectNextServer(void)
+{
+    // This method tries to find the next server info entry in the
+    // Network Data after the current one selected. If found, it
+    // restarts the client with the new server (keeping the retry wait
+    // interval as before).
+
+    Ip6::SockAddr serverSockAddr;
+    bool          selectNext = false;
+
+    serverSockAddr.Clear();
+
+    VerifyOrExit(IsRunning() && mAutoStartModeEnabled && mAutoStartDidSelectServer);
+    VerifyOrExit((mHostInfo.GetState() == kAdding) || (mHostInfo.GetState() == kToAdd));
+
+    // We go through all entries to find the one matching the currently
+    // selected one, then set `selectNext` to `true` so to select the
+    // next one.
+
+    do
+    {
+        NetworkData::Service::DnsSrpAnycast::Info anycastInfo;
+        NetworkData::Service::DnsSrpUnicast::Info unicastInfo;
+        NetworkData::Service::Manager::Iterator   iterator;
+
+        while (Get<NetworkData::Service::Manager>().GetNextDnsSrpAnycastInfo(iterator, anycastInfo) == kErrorNone)
+        {
+            if (selectNext)
+            {
+                serverSockAddr.SetAddress(anycastInfo.mAnycastAddress);
+                serverSockAddr.SetPort(kAnycastServerPort);
+                mServerSequenceNumber           = anycastInfo.mSequenceNumber;
+                mAutoStartIsUsingAnycastAddress = true;
+                ExitNow();
+            }
+
+            if (mAutoStartIsUsingAnycastAddress && (GetServerAddress().GetAddress() == anycastInfo.mAnycastAddress) &&
+                (GetServerAddress().GetPort() == kAnycastServerPort))
+            {
+                selectNext = true;
+            }
+        }
+
+        iterator.Reset();
+
+        while (Get<NetworkData::Service::Manager>().GetNextDnsSrpUnicastInfo(iterator, unicastInfo) == kErrorNone)
+        {
+            if (selectNext)
+            {
+                serverSockAddr                  = unicastInfo.mSockAddr;
+                mAutoStartIsUsingAnycastAddress = false;
+                ExitNow();
+            }
+
+            if (GetServerAddress() == unicastInfo.mSockAddr)
+            {
+                selectNext = true;
+            }
+        }
+
+        // We loop back to handle the case where the current entry
+        // is the last one.
+
+    } while (selectNext);
+
+    // If we reach here it indicates we could not find the entry
+    // associated with currently selected server in the list. This
+    // situation is rather unlikely but can still happen if Network
+    // Data happens to be changed and the entry removed but
+    // the "changed" event from `Notifier` may have not yet been
+    // processed (note that events are emitted from their own
+    // tasklet). In such a case we keep `serverSockAddr` as empty.
+
+exit:
+    if (!serverSockAddr.GetAddress().IsUnspecified())
+    {
+        mHostInfo.SetState(kToAdd);
+        Stop(kRequesterAuto, kKeepRetryInterval);
```

Comment:
Why is the logic ` mHostInfo.SetState(kToAdd);` not part of `Stop` ?

I noticed that Stop will change `kAdding` to `kToRefreshing`. Why doesn't `Stop` change `kAdding` to `kToAdd` directly?

Reply:
Thanks @simonlingoogle .

Yes, In `Stop()`, we do go from `kAdding` to `kToRefresh`.  I added it in first implementation of SRP client with the idea to be safe in the case where client is stopped after we sent an update to server and while waiting for a response from it (there is a a comment in the code which tries to explain it):

```
void Client::Stop(Requester aRequester)
{
    // Change the state of host info and services so that they are
    // added/removed again once the client is started back. In the
    // case of `kAdding`, we intentionally move to `kToRefresh`
    // instead of `kToAdd` since the server may receive our add
    // request and the item may be registered on the server. This
    // ensures that if we are later asked to remove the item, we do
    // notify server.
```

In the new code handling the switch to a new server, we know that we have tried multiple times and waited long enough and received no response from server, so in this case, we manually switch to `kToAdd` (so that we remember we have not yet registered the host info and allow switch to server to happen if things continue to fail.

It's not the best (most clear) approach but it works. I will try to add a comment around `mHostInfo.SetState(kToAdd)` statement to explain this in the code. If you have any suggestions  or ideas around this, please let me know.

Reply:
Thanks for the explanation. That works for me.